### PR TITLE
support openssl ALPN extension

### PR DESCRIPTION
**Pull request by superisaac**
This pr is an implement of ALPN extension, It does't change currenty module too much.
To make alpn work the client needs to call `setAlpnProtocols()` and the server need to call `setAlpnSelectCallback()`, refer to tests/stdlib/thttpclient_ssl_alpn.nim

### CODE REVIEW

File: lib/pure/net.nim [hunk 1]
```diff
@@ -728,6 +728,15 @@ when defineSsl:
     if SSL_set_fd(socket.sslHandle, socket.fd) != 1:
       raiseSSLError()
 
+  proc getAlpnProtocol*(socket: Socket): string =
+    ## Returns the negotiaged ALPN protocols.
+    ## If the ALPN extension is not available or no ALPN candidate
+    ## provided by clients then returns empty string.
+    if not socket.isSsl:
```
Comment (juancarlospaco) [inline lib/pure/net.nim]: The `result` is autoinit to `""` then

```nim
if socket.isSsl: result = socket.sslHandle.getAlpnProtocol()
```

File: lib/pure/asyncnet.nim [hunk 2]
```diff
@@ -772,6 +772,15 @@ when defineSsl:
     of handshakeAsServer:
       sslSetAcceptState(socket.sslHandle)
 
+  proc getAlpnProtocol*(socket: AsyncSocket): string =
+    ## Returns the negotiaged ALPN protocols.
+    ## If the ALPN extension is not available or no ALPN candidate
+    ## provided by clients then returns empty string.
+    if not socket.isSsl:
```
Comment (juancarlospaco) [inline lib/pure/asyncnet.nim]: Same.

File: lib/pure/net.nim (hunk 1 above)
Comment (superisaac) [reply lib/pure/net.nim]: thank you for your comment, the fix commit is pushed as a405eec1

**Pull request closed without merging**